src/metabase/upload.clj

```diff
@@ -603,100 +600,132 @@
   [fields-by-normed-name header]
   ;; Assumes table-cols are unique when normalized
   (let [normalized-field-names (keys fields-by-normed-name)
-        normalized-header (map normalize-column-name header)
-        [extra missing _both] (data/diff (set normalized-header) (set normalized-field-names))]
+        normalized-header      (map normalize-column-name header)
+        [_extra missing _both] (data/diff (set normalized-header) (set normalized-field-names))]
     ;; check for duplicates
     (when (some #(< 1 %) (vals (frequencies normalized-header)))
       (throw (ex-info (tru "The CSV file contains duplicate column names.")
                       {:status-code 422})))
-    (when (or extra missing)
-      (let [error-message (extra-and-missing-error-markdown extra missing)]
+    (when missing
+      (let [error-message (missing-error-markdown missing)]
         (throw (ex-info error-message {:status-code 422}))))))
 
 (defn- matching-or-upgradable? [current-type relaxed-type]
-  (or (= current-type relaxed-type)
+  (or (nil? current-type)
+      (= current-type relaxed-type)
       (when-let [f (allowed-type-upgrades current-type)]
         (f relaxed-type))))
 
 (defn- changed-field->new-type
   "Given some fields and old and new types, filter out fields with unchanged types, then pair with the new types."
   [fields old-types new-types]
-  (let [new-if-changed #(when (not= %1 %2) %2)]
+  (let [new-if-changed #(when (and %1 (not= %1 %2)) %2)]
     (->> (map new-if-changed old-types new-types)
          (map vector fields)
          (filter second)
          (into {}))))
 
+(defn- add-columns! [driver database table field->type & args]
+  (when (seq field->type)
+    (apply driver/add-columns!
+           driver (:id database) (table-identifier table)
+           (m/map-kv (fn [field-or-name column-type]
+                       [(cond
+                          (keyword? field-or-name) field-or-name
+                          (string? field-or-name) (keyword field-or-name)
+                          (map? field-or-name) (keyword (:name field-or-name)))
```

Comment:
where does this complication come from?

Reply:
This comes from my choice to use first-class field entities back when adding the `alter-columns!` method, and then wanting consistency when I added this proxy method. There wasn't much thought put into it, it just felt natural because the fields where in scope.

You're right, it'll be simpler to just require keywords, and there's very little code to change.